### ISSUE TITLE
First Order MAML?

### ISSUE THREAD

**Issue by MurtyShikhar** [facebookresearch/higher#63]
Hi,
Thanks for this very useful piece of software!

I was wondering if there's an easy way to implement first-order MAML with higher. I currently have an implementation of (full) MAML with higher, but wanted to compare it with just first order MAML.

EDIT: It appears that torch.autograd.grad(query_loss, fmodel.parameters()) would give the gradients corresponding to FOMAML and torch.autograd.grad(query_loss, fmodel.parameters(time=0)) can be used to get MAML gradient?

**Comment by Nithin-Holla** [comment]
Yes, I think that's correct! Additionally, you could set the `track_higher_grads` argument to False.

**Comment by egrefen** [comment]
Unless I'm missing something, you actually don't need to use `higher` if you're doing the first order version. Just copy your model, do `k` forward steps, take a derivative, as you've suggested,  of the meta-loss with regard to the latest model weight, and manually map the grads.

As @Nithin-Holla points out, you can use `track_higher_grads` if you want to just wrap your `higher` loop in first-order/forward mode, which is equivalent.

**Comment by rees-c** [comment]
@MurtyShikhar would you mind sharing how you accomplished first order MAML with higher? I tried setting `track_higher_grads` to False as @Nithin-Holla suggested, but found that it resulted in all my meta-learner's gradients being `None` (which did not happen when `track_higher_grads` was True).

Also excuse my ignorance, but what is a clean way to manually map the grads efficiently as in @egrefen's comment? I'm familiar with how to do it for a single layer, but not sure how to do it for the entire network.

Thanks!

**Comment by brando90** [comment]
> Hi, Thanks for this very useful piece of software!
> 
> I was wondering if there's an easy way to implement first-order MAML with higher. I currently have an implementation of (full) MAML with higher, but wanted to compare it with just first order MAML.
> 
> EDIT: It appears that torch.autograd.grad(query_loss, fmodel.parameters()) would give the gradients corresponding to FOMAML and torch.autograd.grad(query_loss, fmodel.parameters(time=0)) can be used to get MAML gradient?

hi @MurtyShikhar how did you end up implemented your fo MAML?

hi @Nithin-Holla I tried setting `track_higher_grads` to False (to run first order maml) but now my gradients are empty! i.e. they are None. Are you suer ethat setting that flag to False is a good idea? Did you try it yourself?

**Comment by brando90** [comment]
@egrefen what is the official way to run first order MAML with higher?

**Comment by kamalojasv181** [comment]
@Nithin-Holla can you confirm if I have understood you correctly or not. MAML to FoMAML can be done just by setting track_higher_grads = False, nothing else?

**Comment by brando90** [comment]
> @Nithin-Holla can you confirm if I have understood you correctly or not. MAML to FoMAML can be done just by setting track_higher_grads = False, nothing else?

@kamalojasv181 @egrefen  whenever I set track_higher_grads=False I get my gradients are Non. I have this check:
```
            assert outer_opt.param_groups[0]['params'][0].grad is not None
```
in my train loop. Is this incorrect or is there something wrong? There should be gradients even if it's FO maml, no? What happens to you when you do: `track_higher_grads=False`?

**Comment by brando90** [comment]
I think this is the solution: https://github.com/facebookresearch/higher/issues/102 to FO

**Comment by brando90** [comment]
duplicate: https://stackoverflow.com/questions/70961541/what-is-the-official-implementation-of-first-order-maml-using-the-higher-pytorch
